File: packet.js
const Type = {
  HEARTBEAT: 0x01,
  TELEMETRY: 0x02,
  IGNITION: 0x03,
  SHUTDOWN: 0x04,
  ACK: 0x05,
  CONF: 0x06,
  INFO: 0x07
}

class Packet {
  constructor() {
    this._type = null
    this.data = ""
  }

  length() {
    return (10 + this.data.length).toString(16).padStart(2, '0')
  }

  set type(type) {
    this._type = type
  }

  get type() {
    return this._type
  }

  set data(data) {
    this._data = data
  }

  get data() {
    return this._data
  }

  encode() {
    var chunk1 = '#' + this.length() + this.type.toString(16).padStart(2, '0') + this.data;
    return chunk1 + this.constructor.fletcher16(chunk1).toString(16).padStart(4, '0') + ';'
  }

  static fletcher16(data) {
    var buf = Buffer.from(data, 'utf8')
    var sum1 = 0xff, sum2 = 0xff;
    var i = 0;
    var len = buf.length;

    while (len) {
      var tlen = len > 20 ? 20 : len;
      len -= tlen;
      do {
        sum2 += sum1 += buf[i++];
      } while (--tlen);
      sum1 = (sum1 & 0xff) + (sum1 >> 8);
      sum2 = (sum2 & 0xff) + (sum2 >> 8);
    }
    /* Second reduction step to reduce sums to 8 bits */
    sum1 = (sum1 & 0xff) + (sum1 >> 8);
    sum2 = (sum2 & 0xff) + (sum2 >> 8);
    return sum2 << 8 | sum1;
  }

  static decode(data) {
    if (data.charAt(0) != '#') {
      throw "Start byte not present, first byte is " + data.charAt(0)
    }
    var length = parseInt(data.charAt(1) + data.charAt(2), 16)
    if (length != data.length) {
      throw "Length does not match, expects " + length + ", got " + data.length
    }
    if (data.charAt(data.length - 1) != ';') {
      throw "End byte not present, last byte is " + data.charAt(data.length - 1)
    }
    if (Object.keys(Type).find(
      key => Type[key] === parseInt(data.charAt(3) + data.charAt(4), 16)
    ) === undefined) {
      throw "Unrecognized packet type: " + parseInt(data.charAt(3) + data.charAt(4), 16)
    }
    var old_checksum = data.slice(length - 5, length - 1)
    var new_checksum = this.fletcher16(data.slice(0, length - 5)).toString(16).padStart(4, '0')
    if (old_checksum != new_checksum) {
      throw "Checksum does not match, expects " + new_checksum + ", got " + old_checksum
    }
    var out = new Packet()
    out.type = parseInt(data.charAt(3) + data.charAt(4), 16)
    out.data = data.slice(5, length - 5)
    return out
  }
}

module.exports = Packet;
module.exports.Type = Type;

File: packet.py
from enum import IntEnum

class Packet():
    class Type(IntEnum):
        HEARTBEAT = 0x01
        TELEMETRY = 0x02
        IGNITION = 0x03
        SHUTDOWN = 0x04
        ACK = 0x05
        CONF = 0x06
        INFO = 0x07

    def __init__(self):
        self.type = ""
        self.data = ""

    @property
    def length(self):
        return self.hexs(10 + len(self.data))

    @property
    def type(self):
        return self._type

    @type.setter
    def type(self, type):
        self._type = type

    @property
    def data(self):
        return self._data

    @data.setter
    def data(self, data):
        self._data = data

    def encode(self):
        chunk1 = '#' + self.length + self.hexs(self.type.value) + self.data
        checksum = self.fletcher16(chunk1)
        return chunk1 + checksum + ';'

    @staticmethod
    def fletcher16(bytestring):
        c0 = 0xff
        c1 = 0xff
        """
        modulus = 255
        bitshift = 8
        bytes_read = 1

        if type(bytestring) != bytes:
            bytestring = str.encode(bytestring)
        bytestring = bytearray(bytestring)

        # an efficient way of doing math.ceil() without frills
        iterations = ((len(bytestring) - 1) // bytes_read) + 1

        # actually updating the checksum by iterating through the bytes
        for i in range(0, iterations):
            start = bytes_read*i
            end = bytes_read*(i+1)

            bytepart = int.from_bytes(bytestring[start:end], byteorder="little")
            c0 = (c0 + bytepart) % modulus
            c1 = (c1 + c0) % modulus

        return hex((c1 << bitshift) + (c0))[2:].zfill(4)
        """
        bytestring = bytearray(str.encode(bytestring))
        i = 0
        length = len(bytestring)
        while length:
            tlen = 20 if length > 20 else length
            length -= tlen
            do = True
            while do:
                c0 += bytestring[i]
                i += 1
                c1 += c0
                tlen -= 1
                do = tlen
            c0 = (c0 & 0xff) + (c0 >> 8)
            c1 = (c1 & 0xff) + (c1 >> 8)
        c0 = (c0 & 0xff) + (c0 >> 8)
        c1 = (c1 & 0xff) + (c1 >> 8)
        return hex(c1 << 8 | c0)[2:].zfill(4)

    @staticmethod
    def decode(raw):
        start = raw[0]
        valid = True
        if start != '#':
            raise DecodeException("Start byte not present, first byte is " + start)
        length = int(raw[1] + raw[2], 16)
        if len(raw) != length:
            raise DecodeException("Length does not match, expects " + str(length) + ", got " + str(len(raw)))
        if raw[len(raw) - 1] != ';':
            raise DecodeException("End byte not present, last byte is " + raw[len(raw) - 1])
        type = int(raw[3] + raw[4], 16)
        checksum = raw[length - 5:length - 1]
        new_checksum = Packet.fletcher16(raw[0:length - 5])
        if checksum != new_checksum:
            raise DecodeException("Checksum does not match, expects " + new_checksum + ", got " + checksum)
        data = "".join(raw[5:length - 5])
        out = Packet()
        try:
            out.type = Packet.Type(type)
        except ValueError as e:
            raise DecodeException("Unrecognized packet type: " + hex(type))
        out.data = data
        return out

    @staticmethod
    def hexs(s):
        h = hex(s)
        if len(h) % 2 == 0:
            return h[2:]
        return '0' + h[2:]

class DecodeException(Exception):
    pass
